va/validation-authority.go

```diff
@@ -447,24 +453,30 @@ func (va *ValidationAuthorityImpl) getCAASet(hostname string) (*CAASet, error) {
 			break
 		}
 
-		// Query CAA records for domain and its alias if it has a CNAME
-		for _, alias := range []bool{false, true} {
-			if alias {
-				target, _, err := va.DNSResolver.LookupCNAME(queryDomain)
-				if err != nil {
-					return nil, err
-				}
-				queryDomain = target
-			}
-			CAAs, _, err := va.DNSResolver.LookupCAA(queryDomain)
+		cnames := 0
+		for {
```

Comment:
Re-reading the RFC I noticed a language difference that might be deliberate:
* "record set returned in response to performing a CAA record query on the label X", vs.
* "a CNAME or DNAME alias record specified at the label X".

Keeping in mind queries are typically performed using resolver libraries and caches (not directly against authoritative servers) it would be reasonable to expect the "record set" to include the results of following a chain of CNAMEs and finding CAAs at the end. So is "CAA, then CNAME" just a natural "normal case first" way to say "follow parents of CNAMEs, not CNAMEs of parents"?
* Look for a CAA record, implicitly incorporating CNAME logic just like all other DNS lookups. If you get one, you're done.
* Otherwise, follow the trail of CNAMEs (if any) before walking up the name hierarchy.

```
a.example.com CNAME b.example.com
b.example.com CNAME c.example.net
c.example.net NXDOMAIN
example.net   CAA   xxxxx
```

This seems more likely to me than the "easier for attackers to add CNAME than to add CAA" explanation, but maybe I just haven't yet imagined how/why that would be easier.

(I also really don't want to believe that a security-through-DNS RFC would deliberately support a validation scenario that depends on breaking the "must not combine CNAME with anything" rule, without even _acknowledging_ it.)

I'm still inclined to return an error if CNAME and CAA both exist for a name. This would have the virtue of failing legitimate validation in the attack scenario you describe, which seems preferable to passing a bogus validation in the converse attack scenario where an attacker can add a CAA but can't remove a CNAME.